free/src/main/scala/cats/free/Cofree.scala

```diff
@@ -0,0 +1,124 @@
+package cats
+package free
+
+/**
+  * A free comonad for some branching functor `S`. Branching is done lazily using Eval.
+  * A tree with data at the branches, as opposed to Free which is a tree with data at the leaves.
+  * Not an instruction set functor made into a program monad as in Free, but an instruction set's outputs as a
+  * functor made into a tree of the possible worlds reachable using the instruction set.
+  */
```

Comment:
Was this derived/based on Scalaz's [Cofree](https://github.com/scalaz/scalaz/blob/series/7.3.x/core/src/main/scala/scalaz/Cofree.scala) or from scratch? If former could you add attribution to Scalaz, similar to what we have [here](https://github.com/typelevel/cats/blob/master/free/src/main/scala/cats/free/FreeT.scala#L14) ?